scripts/gen_regex.py

```diff
@@ -4,89 +4,74 @@
 from pkg_resources import resource_filename
 
 
+CATEGORIES = [unicodedata.category(chr(i)) for i in range(0x110000)]
 DATA_PATH = pathlib.Path(resource_filename('wordfreq', 'data'))
 
 
+def func_to_regex(func):
+    """
+    Given a function that returns True or False for a numerical codepoint,
+    return a regex character class accepting the characters resulting in True.
+    Ranges separated only by unassigned characters are merged for efficiency.
+    """
+    # A list of [start, end (accepted), end (accepted or unassigned)] lists
+    ranges = []
+
+    for i, cat in enumerate(CATEGORIES):
+        if func(i):
+            # If the last range can be extended, do so; else start a new one
+            if ranges and ranges[-1][2] == i - 1:
+                ranges[-1][1] = i
+                ranges[-1][2] = i
+            else:
+                ranges.append([i, i, i])
+        elif cat == 'Cn':
+            # If the last range can be extended, do so
+            if ranges and ranges[-1][2] == i - 1:
+                ranges[-1][2] = i
+
+    return '[%s]' % ''.join(chr(r[0]) + '-' + chr(r[1]) for r in ranges)
+
+
 def cache_regex_from_func(filename, func):
     """
     Generates a regex from a function that accepts a single unicode character,
     and caches it in the data path at filename.
     """
-    with (DATA_PATH / filename).open(mode='w') as file:
-        file.write(func_to_regex(func))
+    with (DATA_PATH / filename).open(mode='w') as f:
+        f.write(func_to_regex(func))
 
 
-def _emoji_char_class():
+def _is_emoji_codepoint(i):
     """
-    Build a regex for emoji substitution.  We create a regex character set
-    (like "[a-cv-z]") matching characters we consider emoji.
+    Report whether a numerical codepoint is (likely) an emoji: a Unicode 'So'
+    character (as future-proofed by the ftfy chardata module) but excluding
+    symbols like © and ™ below U+2600 and the replacement character U+FFFD.
     """
-    cache_regex_from_func(
-        'emoji.txt',
-        lambda c:
-            chardata.CHAR_CLASS_STRING[ord(c)] == '3' and
-            c >= '\u2600' and c != '\ufffd'
-    )
+    return chardata.CHAR_CLASS_STRING[i] == '3' and i >= 0x2600 and i != 0xfffd
 
 
-def _non_punct_class():
+def _is_non_punct_codepoint(i):
     """
-    Builds a regex that matches anything that is not one of the following
-    classes:
+    Report whether a numerical codepoint is not one of the following classes:
     - P: punctuation
     - S: symbols
     - Z: separators
     - C: control characters
-    This will classify symbols, including emoji, as punctuation; callers that
-    want to treat emoji separately should filter them out first.
```

Comment:
(I don't know if that change was deliberate or just an artifact of me typing a particular thing.)